implementations/micrometer-registry-graphite/src/main/java/io/micrometer/graphite/GraphiteMeterRegistry.java

```diff
@@ -71,4 +75,9 @@ public void stop() {
     public void start() {
         this.reporter.start(config.step().getSeconds(), TimeUnit.SECONDS);
     }
+
+    public GraphiteMeterRegistry setNamingConvention(NamingConvention namingConvention) {
```

Comment:
I don't know that we expose the underlying config on any other registries in this way. Is the benefit to just make it easier to find config options?

I believe the config classes were originally created as an effort to make it clear where configuration was set on registries and what was used for metrics.

Reply:
spring boot does not allow you to change the naming convention. The proposed implementation does not meet the requirements in my company (we use dot convention instead of camelCase for storage in graphite). I hope I understood the question correctly?

Reply:
You can modify the meter registry configuration by creating a `MeterConfigurer` bean

```
@Bean
public MeterConfigurer graphiteConfigurer(MeterRegistry reg) {
  if(reg instanceof GraphiteMeterRegistry) {
    reg.config().namingConvention(new namingConvention("myprefix"));
  }
}
```

Reply:
IMHO, the use of "instance of" does not look very elegant. The example with MeterConfigurer is useful, but it seems to me that this method is not very obvious, but it's better to make the prefix in the names of metrics more friendly to programmers. To be honest, you did not convince me. I agree that this example will work, but it looks like a kludge. sorry for my English

Reply:
MeterConfigurer? I not found this class in my spring-boot-2.0.0.m7 application

Reply:
I agree that a type-safe `MeterConfigurer` would be awesome. Not totally sure it's possible in Spring, but would be awesome.

Prefixing as a thing is a very graphite-specific thing unfortunately, so maybe we could get on board with a `GraphiteConfig#prefix()`. On the other hand, we haven't done this historically because the prefix folks add to Graphite metrics tend to be something like application name which really is a _tag_ in a dimensional system. So our advice has been to add a common tag with (following this example) application name, and then override the naming convention to place that tag value in a prefix position in the name. By adding a common tag, the metric is instantly portable to more capable dimensional monitoring systems for which nobody really does prefixing.

That said, we get this request _a lot_ for Graphite users. So much so, that I think it's safe to say the majority of Graphite users want some sort of tag-like prefix on their metrics. So perhaps a `GraphiteConfig#tagAsPrefix(tagKey)` would be even better. Guide the user to adding a common tag that is portable, and imbue the default Graphite naming convention with a knowledge of how to turn the specified tag key into a prefix when it is present.

Reply:
I think @eutkin wants to provide his own `GraphiteMeterRegistry` that is already preconfigured with the naming convention.

```
@Bean
public GraphiteMeterRegistry graphiteRegistry(Clock clock, @Value("${myprefix}") String prefix){
  GraphiteMeterRegistry reg = new GraphiteMeterRegistry(clock)
  reg.setPrefix(prefix)
  return reg;
}
```

That, in fact, seems like more typical Spring Boot autoconfig. Does it make sense to move the `config` class up into the main class? Or are those configs kept separate for the benefit for `@ConfigurationProperties`? I seem to think other boot classes add `@ConfigurationProperties` to the bean in question, instead of it being separate.

Reply:
imho, the whole problem is that we can not use our own naming convention and configure the reporter. Any way to get a link to graphiteMeterRegistry and change the state of this link is a pretty bad decision. Another problem in my opinion is a violation of DI in the class GraphiteMeterRegistry, because it creates its own dependency GraphiteReporter (I believe that the reason in the method is "getDropwizardRegistry ()", that is, it is impossible to get a reference to the internal MetricRegistry). DropwizardMeterRegistry also does not allow you to change your internal meterRegistry, which is not very good for the heirs (as a solution, in the DropwizardMeterRegistry instead of the field reference, use the getter, which you can override). I believe that the problem is somewhat deeper than I thought was originally, and it was a good opportunity to use the external MetricRegistry in the DropwizardMeterRegistry and the heirs, according to the principle of DI

Reply:
I'm having a little trouble following all of that, but I take the point that we should provide an alternate constructor that allows you to provide your own `GraphiteReporter`, and that is I think fair.

Reply:
Maybe so? https://gist.github.com/eutkin/ecfe6f50b7a50a7792d04d83ec09be68 Added setter for MetricRegistry in DropwizardMeterRegistry and other setters in GraphiteMeterRegistry. Not very nice, but it does not break compatibility with spring boot (but there I would make changes, for example, announcements bean MetricExporter for GraphiteMeterRegistry)